bpython/simpleeval.py

```diff
@@ -0,0 +1,234 @@
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+"""simple evaluation of side-effect free code
+
+In order to provide fancy completion, some code can be executed safely.
+
+"""
+
+import ast
+from six import string_types
+from six.moves import builtins
+
+from bpython import line as line_properties
+from bpython._py3compat import py3
+
+class EvaluationError(Exception):
+    """Raised if an exception occurred in safe_eval."""
+
+
+def safe_eval(expr, namespace):
+    """Not all that safe, just catches some errors"""
+    try:
+        return eval(expr, namespace)
+    except (NameError, AttributeError, SyntaxError):
+        # If debugging safe_eval, raise this!
+        # raise
+        raise EvaluationError
+
+
+# This function is under the Python License, Version 2
+# This license requires modifications to the code be reported.
+# Based on ast.literal_eval in Python 2 and Python 3
+# Modifications:
+# * Python 2 and Python 3 versions of the function are combined
+# * checks that objects used as operands of + and - are numbers
+#   instead of checking they are constructed with number literals
+# * new docstring describing different functionality
+# * looks up names from namespace
+# * indexing syntax is allowed
+def simple_eval(node_or_string, namespace=None):
+    """
+    Safely evaluate an expression node or a string containing a Python
+    expression.  The string or node provided may only consist of:
+    * the following Python literal structures: strings, numbers, tuples,
+        lists, and dicts
+    * variable names causing lookups in the passed in namespace or builtins
+    * getitem calls using the [] syntax on objects of the types above
+
+    Like the Python 3 (and unlike the Python 2) literal_eval, unary and binary
+    + and - operations are allowed on all builtin numeric types.
+
+    The optional namespace dict-like ought not to cause side effects on lookup
+    """
+    if namespace is None:
+        namespace = {}
+    if isinstance(node_or_string, string_types):
+        node_or_string = ast.parse(node_or_string, mode='eval')
+    if isinstance(node_or_string, ast.Expression):
+        node_or_string = node_or_string.body
+
+    string_type_nodes = (ast.Str, ast.Bytes) if py3 else (ast.Str,)
+    numeric_types = (int, float, complex) + (() if py3 else (long,))
+
+    # added in Python 3.4
+    if hasattr(ast, 'NameConstant'):
+        name_type_nodes = (ast.Name, ast.NameConstant)
+    else:
+        name_type_nodes = (ast.Name,)
```

Comment:
`string_type_nodes`, `numeric_types`, `name_type_nodes` all look like the could be global.